examples/multigpu_advtrain/model.py

```diff
@@ -0,0 +1,339 @@
+
+    def get_probs(self, x, **kwargs):
+        return self.mlp.fprop(x, **kwargs)
+
+    def fprop(self, x, **kwargs):
+        probs = self.mlp.fprop(x, **kwargs)
+        return {'probs': probs}
+
+    def create_sync_ops(self, host_device):
+        return self.mlp.create_sync_ops(host_device)
+
+
+class Layer(object):
+    def __init__(self, input_shape=None, name=None):
+        self.input_shape = input_shape
+        self.name = name
+        self.params_device = {}
+        self.params_names = None
+        # self.device_name = None  # '/gpu:0'
+        self.device_name = '/gpu:0'
+        self.training = True
+
+    def get_variable(self, name, initializer):
+        v = tf.get_variable(name, shape=initializer.shape,
+                            initializer=lambda shape, dtype,
+                            partition_info:
+                            initializer, trainable=self.training)
+        return v
+
+    def set_input_shape_ngpu(self, new_input_shape, **kwargs):
+        assert self.device_name
+        device_name = self.device_name
+        if self.input_shape is None:
+            # First time setting the input shape
+            self.input_shape = [None] + [int(d) for d in list(new_input_shape)]
+
+        if device_name in self.params_device:
+            # There is a copy of weights on this device
+            self.__dict__.update(self.params_device[device_name])
+            return
+
+        # stop recursion
+        self.params_device[device_name] = {}
+
+        # Initialize weights on this device
+        with tf.device(device_name):
+            keys_before = self.__dict__.keys()
+            self.set_input_shape(self.input_shape, **kwargs)
+            keys_after = self.__dict__.keys()
+            if self.params_names is None:
+                self.params_names = list(set(keys_after) - set(keys_before))
+            params = dict([(k, self.__dict__[k]) for k in self.params_names])
+            self.params_device[device_name] = params
+
+    def create_sync_ops(self, host_device):
+        sync_ops = []
+        host_params = self.params_device[host_device]
+        for device, params in (self.params_device).iteritems():
+            if device == host_device:
+                continue
+            for k in self.params_names:
+                if isinstance(params[k], tf.Variable):
+                    sync_ops += [tf.assign(params[k], host_params[k])]
+        return sync_ops
+
+    def get_output_shape(self):
+        return self.output_shape
+
+    def fprop(self, x, training=False, **kwargs):
+        self.training = training
+        if self.name is None:
+            self.set_input_shape_ngpu(x.shape[1:], **kwargs)
+            return self.fprop_noscope(x, **kwargs)
+        else:
+            with tf.variable_scope(self.name):
+                self.set_input_shape_ngpu(x.shape[1:], **kwargs)
+                return self.fprop_noscope(x, **kwargs)
+
+
+class Linear(Layer):
```

Comment:
similarly, how hard would it be to use the one from the tutorials?

Reply:
Fixed in 714afc0. Changed to LinearnGPU as it has to use the LayernGPUs get_variable.